app/controllers/AddressController.scala

```diff
@@ -793,48 +796,66 @@ class AddressController @Inject()(
               val address =
                 addressDto.toAddress(addressType, journeyData.submittedStartDateDto.fold(LocalDate.now)(_.startDate))
 
-              citizenDetailsService.updateAddress(payeAccount.nino, personDetails.etag, address) map {
+              val originalPostcode = personDetails.address.flatMap(_.postcode).getOrElse("")
 
-                case UpdateAddressBadRequestResponse =>
-                  BadRequest(
-                    views.html.error(
-                      "global.error.BadRequest.title",
-                      Some("global.error.BadRequest.title"),
-                      Some("global.error.BadRequest.message")))
-
-                case UpdateAddressUnexpectedResponse(response) =>
-                  InternalServerError(
-                    views.html.error(
-                      "global.error.InternalServerError500.title",
-                      Some("global.error.InternalServerError500.title"),
-                      Some("global.error.InternalServerError500.message")))
-
-                case UpdateAddressErrorResponse(cause) =>
-                  InternalServerError(
-                    views.html.error(
-                      "global.error.InternalServerError500.title",
-                      Some("global.error.InternalServerError500.title"),
-                      Some("global.error.InternalServerError500.message")))
-
-                case UpdateAddressSuccessResponse =>
+              addressMovedService.moved(originalPostcode, address.postcode.getOrElse("")).flatMap { addressChanged =>
+                def successResponseBlock(): Result = {
+                  val originalAddressDto: Option[AddressDto] =
+                    journeyData.selectedAddressRecord.map(AddressDto.fromAddressRecord)
                   handleAddressChangeAuditing(originalAddressDto, addressDto, personDetails, addressType)
-                  clearCache() //This clears ENTIRE session cache, no way to target individual keys
-                  Ok(views.html.personaldetails.updateAddressConfirmation(typ, false, None))
+                  clearCache()
+
+                  Ok(views.html.personaldetails
+                    .updateAddressConfirmation(typ, false, None, addressMovedService.toMessageKey(addressChanged)))
+                }
+
+                updateAddress(journeyData, addressType, payeAccount, personDetails, address, successResponseBlock)
               }
             }
           }
         }
       }
     }
 
+  private def updateAddress(
+    journeyData: AddressJourneyData,
+    addressType: String,
+    payeAccount: PayeAccount,
+    personDetails: PersonDetails,
+    address: Address,
+    successResponseBlock: () => Result)(implicit pertaxContext: PertaxContext): Future[Result] =
+    citizenDetailsService.updateAddress(payeAccount.nino, personDetails.etag, address) map {
+
+      case UpdateAddressBadRequestResponse =>
+        BadRequest(
+          views.html.error(
```

Comment:
These views and messages seem very generic, is there a better place for them/

Reply:
Moved these and the UpdateAddressResponse behaviour, along with the tests